### PR TITLE
[Broker Clean up] Decouple broker from transport package

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Change BrokerRequestHandler to an interface
Move all transport package related classes into ConnectionPoolBrokerRequestHandler implementation
Enhancing broker query log to include servers and exceptions info

Query log example:
2018/08/22 16:53:53.816 INFO [com.linkedin.pinot.broker.requesthandler.ConnectionPoolBrokerRequestHandler] [] Request 7 ScatterGatherStats: <hostName>_O=0,0,0,62
2018/08/22 16:53:53.816 INFO [com.linkedin.pinot.broker.requesthandler.BaseBrokerRequestHandler] [] Request 7:myTable_OFFLINE, time:100ms, docs:10000/44673408, entries:0/80000, servers:1/1, exceptions:0, query:select * from myTable limit 10000

Explaination: first line is the ScatterGatherStats, second line includes tableName, queryTime, numDocsScanned/numTotalDocs, numEntriesScannedIn/PostFilter, serverResponded/Queried, numExceptions, query

### CODE REVIEW

**Comment by Jackie-Jiang** [comment]
@kishoreg Yes, this is for the broker async routing (or any new routing strategy). Will create an issue for this.

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java [hunk 1]
```diff
@@ -107,81 +65,29 @@
...
-    SegmentZKMetadataPrunerService brokerPrunerService = new SegmentZKMetadataPrunerService(prunerNames);
-
-    // Setup Broker Request Handler
-    ReduceServiceRegistry reduceServiceRegistry = buildReduceServiceRegistry();
-    _accessControlFactory = AccessControlFactory.loadFactory(_config.subset(BROKER_ACCESS_CONTROL_PREFIX));
-    _requestHandler = new BrokerRequestHandler(_routingTable, _timeBoundaryService, _scatterGather,
-        reduceServiceRegistry, brokerPrunerService, _brokerMetrics, _config, _accessControlFactory,
-        _tableQueryQuotaManager);
+    _brokerRequestHandler =
+        new ConnectionPoolBrokerRequestHandler(_config, _routingTable, _timeBoundaryService, _accessControlFactory,
```
Comment (snleee) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java]: What is the plan for the migration after you implement async request handler? After your refactoring, we will have multiple implementations of broker request handlers. We may need to consider to add the configuration and use the factory to initialize broker request handler (and use async one by default).

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+      // No table matches the request
+      LOGGER.info("No table matches the name: {}", tableName);
+      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.RESOURCE_MISSING_EXCEPTIONS, 1);
+      return BrokerResponseNative.NO_TABLE_RESULT;
+    }
+
+    // Validate QPS quota
+    if (!_tableQueryQuotaManager.acquire(tableName)) {
+      String errorMessage = String.format("Request %d exceeds query quota for table: %s", requestId, tableName);
+      LOGGER.info(errorMessage);
```
Comment (snleee) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Should it be `LOGGER.error`?

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+      _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.QUERY_QUOTA_EXCEEDED, 1);
+      return new BrokerResponseNative(QueryException.getException(QueryException.QUOTA_EXCEEDED_ERROR, errorMessage));
+    }
+
+    // Validate the request
+    try {
+      validateRequest(brokerRequest);
+    } catch (Exception e) {
+      LOGGER.info("Caught exception while validating request {}: {}, {}", requestId, query, e.getMessage());
```
Comment (snleee) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Same here (`LOGGER.error`?)

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 2 above)
Comment (mcvsubbu) [reply pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: No, this is not an error in Pinot. It is a user error. So we should log it as INFO and return 4xx to user.

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    _accessControlFactory = accessControlFactory;
+    _tableQueryQuotaManager = tableQueryQuotaManager;
+    _brokerMetrics = brokerMetrics;
+
+    _brokerId = config.getString(CONFIG_OF_BROKER_ID, getDefaultBrokerId());
+    _brokerTimeoutMs = config.getLong(CONFIG_OF_BROKER_TIMEOUT_MS, DEFAULT_BROKER_TIMEOUT_MS);
+    _queryResponseLimit = config.getInt(CONFIG_OF_BROKER_QUERY_RESPONSE_LIMIT, DEFAULT_BROKER_QUERY_RESPONSE_LIMIT);
+    _queryLogLength = config.getInt(CONFIG_OF_BROKER_QUERY_LOG_LENGTH, DEFAULT_BROKER_QUERY_LOG_LENGTH);
+
+    LOGGER.info("Broker Id: {}", _brokerId);
```
Comment (mcvsubbu) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Combine these into a single log message?

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 3 above)
Comment (mcvsubbu) [reply pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: same

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/com/linkedin/pinot/broker/broker/BrokerServerBuilder.java]: That will be in the PR of async request handler

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Done

**Pull request merged**